compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/CFGNode.kt

```diff
@@ -19,25 +19,34 @@ import org.jetbrains.kotlin.utils.addToStdlib.runIf
 
 sealed class CFGNode<out E : FirElement>(val owner: ControlFlowGraph, val level: Int, private val id: Int) {
     companion object {
-        internal fun addEdge(from: CFGNode<*>, to: CFGNode<*>, kind: EdgeKind, propagateDeadness: Boolean) {
+        internal fun addEdge(from: CFGNode<*>, to: CFGNode<*>, kind: EdgeKind, propagateDeadness: Boolean, label: String? = null) {
             from._followingNodes += to
             to._previousNodes += from
-            addJustKindEdge(from, to, kind, propagateDeadness, edgeExists = false)
+            addJustKindEdge(from, to, kind, propagateDeadness, edgeExists = false, label = label)
         }
 
-        internal fun addJustKindEdge(from: CFGNode<*>, to: CFGNode<*>, kind: EdgeKind, propagateDeadness: Boolean) {
-            addJustKindEdge(from, to, kind, propagateDeadness, edgeExists = true)
+        internal fun addJustKindEdge(from: CFGNode<*>, to: CFGNode<*>, kind: EdgeKind, propagateDeadness: Boolean, label: String? = null) {
+            addJustKindEdge(from, to, kind, propagateDeadness, edgeExists = true, label = label)
         }
 
-        private fun addJustKindEdge(from: CFGNode<*>, to: CFGNode<*>, kind: EdgeKind, propagateDeadness: Boolean, edgeExists: Boolean) {
-            if (kind != EdgeKind.Forward) {
-                val fromToKind = from._outgoingEdges[to] ?: runIf(edgeExists) { EdgeKind.Forward }
+        private fun addJustKindEdge(
+            from: CFGNode<*>,
+            to: CFGNode<*>,
+            kind: EdgeKind,
+            propagateDeadness: Boolean,
+            edgeExists: Boolean,
+            label: String? = null
+        ) {
+            // It's hard to define label merging, hence overwritten with the latest one.
+            // One day, if we allow multiple edges between nodes with different labels, we won't even need kind merging.
+            if (kind != EdgeKind.Forward || label != null) {
+                val fromToKind = from._outgoingEdges[to]?.kind ?: runIf(edgeExists) { EdgeKind.Forward }
                 merge(kind, fromToKind)?.let {
-                    from._outgoingEdges[to] = it
+                    from._outgoingEdges[to] = Edge(it, label)
```

Comment:
Since almost all edges in user code without labels, there will be a lot of same objects of `Edge` with four different kinds and `null` as label which is a waste of memory. One of possible solutions is creating four static instances of `Edge` and using them if `label == null`

Reply:
Good idea, thanks for the suggestion. I assume four kinds you mentioned are these: https://github.com/JetBrains/kotlin/blob/master/compiler/tests-common/tests/org/jetbrains/kotlin/fir/AbstractFirDiagnosticTest.kt#L324, but I thought it won't hurt to cover all edge kinds when there is no label. Added a creating util that returns static ones appropriately.